actors/builtin/miner/miner_actor.go

```diff
@@ -487,6 +467,186 @@ func (a Actor) SubmitWindowedPoSt(rt Runtime, params *SubmitWindowedPoStParams)
 	return nil
 }
 
+type ChallengeWindowedPoStParams struct {
+	Deadline   uint64
+	ProofIndex uint64 // only one is allowed at a time to avoid loading too many sector infos.
+}
+
+func (a Actor) ChallengeWindowedPoSt(rt Runtime, params *ChallengeWindowedPoStParams) *abi.EmptyValue {
+	rt.ValidateImmediateCallerAcceptAny()
+
+	if params.Deadline >= WPoStPeriodDeadlines {
+		rt.Abortf(exitcode.ErrIllegalArgument, "invalid deadline %d of %d", params.Deadline, WPoStPeriodDeadlines)
+	}
+
+	currEpoch := rt.CurrEpoch()
+
+	// TODO: maybe stash this in the deadline for better accuracy?
+	epochReward := requestCurrentEpochBlockReward(rt)
+	pwrTotal := requestCurrentTotalPower(rt)
+
+	toBurn := abi.NewTokenAmount(0)
+	pledgeDelta := abi.NewTokenAmount(0)
+	powerDelta := NewPowerPairZero()
+	var st State
+	rt.StateTransaction(&st, func() {
+		if st.CurrentDeadline == params.Deadline {
+			// TODO: blank out the previous deadline as well?
+			rt.Abortf(exitcode.ErrForbidden, "cannot challenge a window post for an open deadline")
+		}
+
+		info := getMinerInfo(rt, &st)
+		faultyPower := NewPowerPairZero()
+		store := adt.AsStore(rt)
+
+		// Check proof
+		// TODO: move into deadline state function.
+		{
+			// Load the target state.
+			deadlines, err := st.LoadDeadlines(store)
+			builtin.RequireNoErr(rt, err, exitcode.ErrIllegalState, "failed to load deadlines")
+
+			dl, err := deadlines.LoadDeadline(store, params.Deadline)
+			builtin.RequireNoErr(rt, err, exitcode.ErrIllegalState, "failed to load deadline")
+
+			proofs, err := adt.AsArray(store, dl.ProofsSnapshot)
```

Comment:
Snapshot vs current state vars will be easily confused, so I suggest naming quite explicitly:
- `dlCurrent`
- `proofsSnapshot`
- etc